Detectors/TRD/calibration/src/TrackBasedCalib.cxx

```diff
@@ -77,6 +90,78 @@ void TrackBasedCalib::calculateAngResHistos()
   // mAngResHistos.print();
 }
 
+int TrackBasedCalib::filldEdx(gsl::span<const TrackTRD>& tracks, bool isTPCTRD)
+{
+  auto& params = TRDCalibParams::Instance();
+  int nTracksSuccess = 0;
+  for (const auto& trkIn : tracks) {
+
+    if (trkIn.getNtracklets() < params.nTrackletsMinGainCalib) {
+      continue;
+    }
+
+    if (trkIn.getP() < params.pMin || trkIn.getP() > params.pMax) {
+      continue;
+    }
+
+    auto id = trkIn.getRefGlobalTrackId();
+    float dEdxTPC;
+    if (isTPCTRD) {
+      dEdxTPC = mTracksTPC[id].getdEdx().dEdxTotTPC;
+    } else {
+      dEdxTPC = mTracksTPC[mTracksITSTPC[id].getRefTPC()].getdEdx().dEdxTotTPC;
+    }
+    if (dEdxTPC < params.dEdxTPCMin || dEdxTPC > params.dEdxTPCMax) {
+      continue;
+    }
+
+    if (std::isnan(trkIn.getSnp())) {
+      LOG(alarm) << "Track with invalid parameters found: " << trkIn.getRefGlobalTrackId();
+      continue;
+    }
+
+    auto trkWork = trkIn;
+
+    for (int iLayer = NLAYER - 1; iLayer >= 0; --iLayer) {
+      if (trkWork.getTrackletIndex(iLayer) == -1) {
+        continue;
+      }
+      if (mNoiseCalib && mNoiseCalib->isTrackletFromNoisyMCM(mTrackletsRaw[trkWork.getTrackletIndex(iLayer)])) {
+        // ignore tracklets which originate from noisy MCMs
+        continue;
+      }
+      if (propagateAndUpdate(trkWork, iLayer, false)) {
+        break;
+      }
+
+      float trkPhi = o2::math_utils::asin(trkWork.getSnp());
+      float trkLength = sqrt(1 + tan(trkPhi) * tan(trkPhi) + trkWork.getTgl() * trkWork.getTgl());
```

Comment:
the trigonometric functions and especially the `asin` operation are rather slow, so you should check for an alternative in places which are called often like here for every track. Instead of `tan (phi)^2` you can substitute `sin (phi)^2 / (1 - sin(phi)^2)` so you can use the parameter stored in the track directly